File: {{ cookiecutter.repo_name }}/{{ cookiecutter.module_name }}/data/cross_language.py
"""Cross-language data exchange example.

Demonstrates using Arrow/Feather format for sharing data
between Python and Julia (or other languages).
"""
import pandas as pd
from loguru import logger

from {{ cookiecutter.module_name }}.config import INTERIM_DATA_DIR, PROCESSED_DATA_DIR

# === Configuration ===
# Data to send to Julia
EXPORT_FILE = INTERIM_DATA_DIR / "for_julia.feather"

# Data received from Julia
IMPORT_FILE = INTERIM_DATA_DIR / "from_julia.feather"

# Final output
OUTPUT_FILE = PROCESSED_DATA_DIR / "final_results.parquet"

# === Export to Julia ===
logger.info("Preparing data for Julia...")

# Example: Create some data to process in Julia
df_export = pd.DataFrame(
    {"x": range(100), "y": [i**2 for i in range(100)], "category": ["A", "B"] * 50}
)

logger.info(f"Exporting to Feather: {EXPORT_FILE}")
df_export.to_feather(EXPORT_FILE)

logger.info("Data exported. Now run your Julia script to process this file.")
logger.info(
    "Julia example: using Arrow, DataFrames; df = DataFrame(Arrow.Table('for_julia.feather'))"
)

# === Import from Julia ===
# Assuming Julia has processed and saved results back
if IMPORT_FILE.exists():
    logger.info(f"Importing results from Julia: {IMPORT_FILE}")
    df_result = pd.read_feather(IMPORT_FILE)

    # Convert to Parquet for final storage
    logger.info(f"Saving final results: {OUTPUT_FILE}")
    df_result.to_parquet(OUTPUT_FILE)

    logger.success("Cross-language workflow complete.")
else:
    logger.warning(f"Julia output not found: {IMPORT_FILE}")
    logger.info("Waiting for Julia to process the data...")

File: {{ cookiecutter.repo_name }}/src/data/cross_language.jl
"""
Cross-language data exchange example.
Mirrors {{ cookiecutter.module_name }}/data/cross_language.py
Demonstrates using Arrow/Feather for sharing data between Python and Julia.
"""

using DataFrames, Arrow
using Logging

include("../config.jl")

# === Configuration ===
const EXPORT_FILE = joinpath(INTERIM_DATA_DIR, "julia_to_python.arrow")
const IMPORT_FILE = joinpath(INTERIM_DATA_DIR, "python_to_julia.arrow")
const OUTPUT_FILE = joinpath(PROCESSED_DATA_DIR, "final_results.arrow")

# === Export to Python ===
@info "Preparing data for Python..."

df_export = DataFrame(x=0:99, y=[i^2 for i in 0:99], category=repeat(["A","B"], 50))

@info "Exporting to Arrow: $EXPORT_FILE"
Arrow.write(EXPORT_FILE, df_export)

@info "Data exported. Python can read with: pyarrow.feather.read_table('julia_to_python.arrow')"

# === Import from Python ===
if isfile(IMPORT_FILE)
    @info "Importing results from Python: $IMPORT_FILE"
    df_result = DataFrame(Arrow.Table(IMPORT_FILE))
    @info "Saving final results: $OUTPUT_FILE"
    Arrow.write(OUTPUT_FILE, df_result)
    @info "Cross-language workflow complete."
else
    @warn "Python output not found: $IMPORT_FILE"
    @info "Waiting for Python to process the data..."
end
